### PR TITLE
[shopsys] model service layer removal (phase 5)

### PR DESCRIPTION
**Pull request by simara-svatopluk**
Model service layer doesn't make sense too much. Should I put business logic into entities, services or both? We decided that business logic belongs to entities, we can get rid of model services and simplify the model.

This is a quite big change, so we split it into several phases.

| Q             | A
| ------------- | ---
|Description, reason for the PR| 
|New feature| No
|BC breaks| Yes
|Fixes issues|  
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: docs/upgrade/services_removal.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ Following instructions tell you which method was moved where, so you can upgrade
 
 If you use these methods, change their calling appropriately:
 - `CustomerService::edit(User $user, UserData $userData)`  
-  -> `User::edit(UserData $userData, CustomerPasswordService $customerPasswordService)`
+  -> `edit(UserData $userData, EncoderFactoryInterface $encoderFactory)`
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: You removed "User", probably by accident :slightly_smiling_face:

File: docs/upgrade/services_removal.md [hunk 2]
```diff
@@ -58,6 +58,34 @@ If you use these methods, change their calling appropriately:
   -> `RegistrationMail::createMessage(MailTemplate $mailTemplate, $user)`
 - `OrderMailService::getMessageDataByOrder(Order $order, MailTemplate $mailTemplate)`  
   -> `OrderMail::createMessage(MailTemplate $mailTemplate, $order)`
+- `CustomerPasswordService::resetPassword(User $user)`  
+  -> `User::resetPassword(HashGenerator $hashGenerator)`
+- `CustomerPasswordService::isResetPasswordHashValid(User $user, $hash)`  
+  -> `User::isResetPasswordHashValid(?string $hash): bool`
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: the return type is not mentioned anywhere else in this document

File: docs/upgrade/services_removal.md [hunk 3]
```diff
@@ -58,6 +58,34 @@ If you use these methods, change their calling appropriately:
...
+- `CustomerPasswordService::changePassword(User $user, $password)`  
+  -> `User::changePassword(EncoderFactoryInterface $encoderFactory, $password)`
+- `CustomerPasswordService::setNewPassword(User $user, $hash, $newPassword)`  
+  -> `User::setNewPassword(EncoderFactoryInterface $encoderFactory, ?string $hash, string $newPassword)`
+- `ProductVariantService::checkProductIsNotMainVariant(Product $product)`  
+  -> `Product::checkIsNotMainVariant()`
+- `ProductVariantService::refreshProductVariants(Product $mainProduct, array $currentVariants)`  
+  -> `Product::refreshProductVariants(array $currentVariants)`
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: maybe "refreshVariants" would be better, however, you would need to change it in e118616801800fdb776aaac0f729f7da2e513932 as well. I do not insist on the renaming, it is up to you :wink:

File: docs/upgrade/services_removal.md [hunk 4]
```diff
@@ -99,6 +144,7 @@ Following functions visibility was changed to `protected` as there is no need to
 Follow also additional upgrade instructions:
 - Change return type of `DeliveryAddressFactory::create()` to `?DeliveryAddress` as it now returns `null` when `addressFilled` is `false`
 - Change usage of `ProductListOrderingModeService::` constants to `ProductListOrderingConfig::` as they were moved to the Config class
+- Change callin of `OrderMail::getMailTemplateNameByStatus()` to static calling as the method is static now
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: typo - "callin**g**"

File: packages/framework/src/Component/Image/ImageFacade.php [hunk 5]
```diff
@@ -306,4 +306,16 @@ public function copyImages($sourceEntity, $targetEntity)
         }
         $this->em->flush($targetImages);
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Image\Image[] $orderedImages
+     */
+    private function setImagePositionsByOrder($orderedImages)
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Image/ImageFacade.php]: should be protected

File: docs/upgrade/services_removal.md [hunk 6]
```diff
@@ -75,6 +107,12 @@ Following methods have been removed:
 - `Administrator::getLimitByGridId()` as it was not used anywhere
 - `FlagService::create()`, use `FlagFactory::create()` instead
 - `FlagService::edit()`, use `Flag::edit()` instead
+- `User::setResetPasswordHash()`, use `User::resetPassword()` instead
+- `User::getResetPasswordHashValidThrough()` as it was not used anywhere
+- `ImageService::setImagePositionsByOrder()` as it was moved to private method of `ImageFacade`
+- `ImageService::createImage()`, use `ImageFactory::create()` instead
+- `FriendlyUrlService::getAbsoluteUrlByDomainConfigAndSlug()` as it was moved to private method of `SitemapListener`
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: the method should be protected, not private

File: packages/framework/src/Component/Image/ImageFacade.php [hunk 7]
```diff
@@ -50,6 +50,11 @@ class ImageFacade
      */
     protected $imageUrlPrefix;
 
+    /**
+     * @var \Shopsys\FrameworkBundle\Component\Image\ImageFactoryInterface
+     */
+    private $imageFactory;
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Image/ImageFacade.php]: should be protected

File: packages/framework/tests/Unit/Component/Image/ImageFactoryTest.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+use Shopsys\FrameworkBundle\Component\FileUpload\FileUpload;
+use Shopsys\FrameworkBundle\Component\Image\Config\ImageEntityConfig;
+use Shopsys\FrameworkBundle\Component\Image\Image;
+use Shopsys\FrameworkBundle\Component\Image\ImageFactory;
+use Shopsys\FrameworkBundle\Component\Image\Processing\ImageProcessor;
+use Symfony\Component\Filesystem\Filesystem;
+
+class ImageFactoryTest extends TestCase
+{
+    public function testCreateImage()
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Image/ImageFactoryTest.php]: should be `testCreate()` :slightly_smiling_face:

File: packages/framework/tests/Unit/Component/Image/ImageFactoryTest.php [hunk 9]
```diff
@@ -17,6 +17,46 @@
...
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $imageFactory = new ImageFactory($imageProcessorMock, $this->getFileUpload());
+
+        $this->expectException(\Shopsys\FrameworkBundle\Component\Image\Exception\EntityMultipleImageException::class);
+        $imageFactory->createMultiple($imageEntityConfig, 1, 'type', []);
+    }
+
+    public function testGetUploadedImages()
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Image/ImageFactoryTest.php]: should be `testCreateMultiple()`

File: packages/framework/src/Model/Sitemap/SitemapListener.php [hunk 10]
```diff
@@ -128,4 +119,14 @@ private function addHomepageUrl(
         $urlConcrete = new UrlConcrete($absoluteUrl, null, null, $elementPriority);
         $generator->addUrl($urlConcrete, $section);
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig $domainConfig
+     * @param string $slug
+     * @return string
+     */
+    private function getAbsoluteUrlByDomainConfigAndSlug(DomainConfig $domainConfig, $slug)
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Sitemap/SitemapListener.php]: rather `protected` :innocent:

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php [hunk 11]
```diff
@@ -19,4 +21,29 @@ public function create(
...
+     * @param int|null $indexPostfix
+     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrl|null
+     */
+    public function createIfValid(
+        string $routeName,
+        int $entityId,
+        string $entityName,
+        int $domainId,
+        int $indexPostfix = null
+    ):?FriendlyUrl {
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php]: there should be a space between ":" and "?", right? We should have a fixer for that :slightly_smiling_face:

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php [hunk 12]
```diff
@@ -46,4 +60,35 @@ public function createIfValid(
 
         return $this->create($routeName, $entityId, $domainId, $slug);
     }
+
+    /**
+     * @param string $routeName
+     * @param int $entityId
+     * @param string[] $namesByLocale
+     * @param \Shopsys\FrameworkBundle\Component\Domain\Domain $domain
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php]: this does not belong here

File: packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlFactoryTest.php [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+use PHPUnit\Framework\TestCase;
+use Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig;
+use Shopsys\FrameworkBundle\Component\Domain\Domain;
+use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlFactory;
+use Shopsys\FrameworkBundle\Component\Setting\Setting;
+
+class FriendlyUrlFactoryTest extends TestCase
+{
+    public function testCreateFriendlyUrls()
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlFactoryTest.php]: should be `testCreateForAllDomains`

File: packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlTest.php [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+use PHPUnit\Framework\TestCase;
+use Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig;
+use Shopsys\FrameworkBundle\Component\Domain\Domain;
+use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrl;
+use Shopsys\FrameworkBundle\Component\Setting\Setting;
+
+class FriendlyUrlTest extends TestCase
+{
+    public function testGetAbsoluteUrlByFriendlyUrl()
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlTest.php]: should be `testGetAbsoluteUrl()`

File: packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlUniqueResultFactoryTest.php [hunk 15]
```diff
@@ -7,25 +7,25 @@
...
 use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlFactory;
-use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlService;
+use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlUniqueResultFactory;
 use Shopsys\FrameworkBundle\Component\Setting\Setting;
 
-class FriendlyUrlServiceTest extends TestCase
+class FriendlyUrlUniqueResultFactoryTest extends TestCase
 {
-    public function testGetFriendlyUrlUniqueResultNewUnique()
+    public function testCreateFriendlyUrlUniqueResultNewUnique()
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlUniqueResultFactoryTest.php]: should be `testCreateNewUnique()`

File: packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlUniqueResultFactoryTest.php [hunk 16]
```diff
@@ -36,23 +36,23 @@ public function testGetFriendlyUrlUniqueResultNewUnique()
         $this->assertSame($friendlyUrl, $friendlyUrlUniqueResult->getFriendlyUrlForPersist());
     }
 
-    public function testGetFriendlyUrlUniqueResultOldUnique()
+    public function testCreateFriendlyUrlUniqueResultOldUnique()
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlUniqueResultFactoryTest.php]: should be `testCreateOldUnique()`

File: packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlUniqueResultFactoryTest.php [hunk 17]
```diff
@@ -63,23 +63,23 @@ public function testGetFriendlyUrlUniqueResultOldUnique()
         $this->assertNull($friendlyUrlUniqueResult->getFriendlyUrlForPersist());
     }
 
-    public function testGetFriendlyUrlUniqueResultNotUnique()
+    public function testCreateFriendlyUrlUniqueResultNotUnique()
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Router/FriendlyUrl/FriendlyUrlUniqueResultFactoryTest.php]: should be `testCreateNotUnique()`

File: project-base/tests/ShopBundle/Functional/Model/Order/Mail/OrderMailTest.php [hunk 18]
```diff
@@ -23,35 +23,6 @@ class OrderMailTest extends FunctionalTestCase
 {
     public function testGetMailTemplateNameByStatus()
     {
-        $routerMock = $this->getMockBuilder(RouterInterface::class)->setMethods(['generate'])->getMockForAbstractClass();
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Functional/Model/Order/Mail/OrderMailTest.php]: :ok_hand:

File: packages/framework/src/Model/Mail/MailTemplateFacade.php [hunk 19]
```diff
@@ -99,12 +100,27 @@ public function getOrderStatusMailTemplatesIndexedByOrderStatusId($domainId)
...
             $mailTemplates
         );
     }
 
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Order\Status\OrderStatus[] $orderStatuses
+     * @param \Shopsys\FrameworkBundle\Model\Mail\MailTemplate[] $mailTemplates
+     * @return \Shopsys\FrameworkBundle\Model\Mail\MailTemplate[]
+     */
+    private function getFilteredOrderStatusMailTemplatesIndexedByOrderStatusId(array $orderStatuses, array $mailTemplates)
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Mail/MailTemplateFacade.php]: protected :slightly_smiling_face:

File: packages/framework/src/Component/Image/ImageFactory.php [hunk 20]
```diff
@@ -17,6 +41,36 @@ public function create(
         ?string $type,
         ?string $temporaryFilename
     ): Image {
-        return new Image($entityName, $entityId, $type, $temporaryFilename);
+        $temporaryFilePath = $this->fileUpload->getTemporaryFilepath($temporaryFilename);
```
Comment (boris-brtan) [inline packages/framework/src/Component/Image/ImageFactory.php]: ?string $temporaryFilename $temporaryFilePath = $this->fileUpload->getTemporaryFilepath($temporaryFilename);
can we discuss this change a bit ?

Comment (simara-svatopluk) [reply packages/framework/src/Component/Image/ImageFactory.php]: The commit message is wrong, it should have been "from Service to Factory" (not Facade) https://github.com/shopsys/shopsys/commit/830f51d7bf07e04b2430b5d92775d8eccb9ffe36

Click into the commit and find changes in `ImageService`, this logic was moved to `ImageFactory`

Comment (vitek-rostislav) [reply packages/framework/src/Component/Image/ImageFactory.php]: I will take a look at it... :wink:

**Pull request merged**